File: LiBattleship/ClientApp/models/Field.ts
export class Field {
    map: Array<Array<number>>;
    ships: any;

    constructor(map?: Array<Array<number>>) {
        this.ships = { '1': 0, '2': 0, '3': 0, '4': 0 };
        if (map) {
            this.map = map;
            this.recalcShips();
        } else {
            this.map = [];
            for (let i = 0; i < 10; i++) {
                this.map.push([0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);
            }
        }
    }

    public makeAction(x: number, y: number): boolean {
        if (this.map[x][y] == 0 && !this.isValidMove(x, y)) return false;
        let nearestPoints = this.getNearestPoints(x, y);

        if (this.map[x][y] == 0) {
            nearestPoints.map((point) => this.map[x + point.dx][y + point.dy])
                .forEach((ship) => this.ships[ship]--);
            this.setPointWeight(x, y, 1);
            this.ships[this.map[x][y]]++;
        } else {
            this.ships[this.map[x][y]]--;
            this.setPointWeight(x, y, 0);
            this.getNearestPoints(x, y).map((point) => this.map[x + point.dx][y + point.dy])
                .forEach((ship) => this.ships[ship]++);
        }
        console.log(this.ships);
        return true;
    }

    public getState(x: number, y: number): boolean {
        return this.map[x][y] > 0;
    }

    private setPointWeight(x: number, y: number, weight: number) {
        this.map[x][y] = weight;
        if (this.map[x][y] > 0) this.setRegionWeight(x, y, this.getRegionWeight(x, y));
        else {
            this.getNearestPoints(x, y).forEach((point) => {
                if (point.dx != 0 || point.dy != 0) this.setRegionWeight(x + point.dx, y + point.dy, this.getRegionWeight(x + point.dx, y + point.dy));
            });
        }
    }

    private getRegionWeight(x: number, y: number): number {
        let sum = 0;
        let nearestPoints = this.getNearestPoints(x, y);
        nearestPoints.forEach((point) => {
            let i = x, j = y;
            while (true) {
                sum++;
                if (i + point.dx < this.map.length && i + point.dx >= 0
                    && j + point.dy < this.map.length && j + point.dy >= 0 && this.map[i + point.dx][j + point.dy] > 0) {
                    i = i + point.dx;
                    j = j + point.dy;
                }
                else break;
                if (point.dx == 0 && point.dy == 0) break;
            }
        })

        return sum - nearestPoints.length + 1;
    }

    private setRegionWeight(x: number, y: number, weight: number) {
        this.getNearestPoints(x, y).forEach((point) => {
            let i = x, j = y;
            while (true) {
                this.map[i][j] = weight;
                if (i + point.dx < this.map.length && i + point.dx >= 0 && j + point.dy < this.map.length
                    && j + point.dy >= 0 && this.map[i + point.dx][j + point.dy] > 0) {
                    i = i + point.dx;
                    j = j + point.dy;
                }
                else break;
                if (point.dx == 0 && point.dy == 0) break;
            }
        })
    }

    private getNearestPoints(x: number, y: number): Array<{ dx: number, dy: number }> {
        let result = [];
        if (this.map[x][y] > 0) { result.push({ dx: 0, dy: 0 }) }
        if (x > 0 && this.map[x - 1][y] > 0) { result.push({ dx: -1, dy: 0 }) }
        if (y < this.map.length - 1 && this.map[x][y + 1] > 0) { result.push({ dx: 0, dy: 1 }) }
        if (y > 0 && this.map[x][y - 1] > 0) { result.push({ dx: 0, dy: -1 }) }
        if (x < this.map.length - 1 && this.map[x + 1][y] > 0) { result.push({ dx: 1, dy: 0 }) }
        return result;
    }

    private isValidMove(i: number, j: number): boolean {
        let result = true;
        if (i > 0 && j > 0 && this.map[i - 1][j - 1]) return false;
        if (i > 0 && j < this.map.length - 1 && this.map[i - 1][j + 1]) return false;
        if (j > 0 && i < this.map.length - 1 && this.map[i + 1][j - 1]) return false;
        if (j < this.map.length - 1 && i < this.map.length - 1 && this.map[i + 1][j + 1]) return false;
        let probableShip = this.getNearestPoints(i, j).map((point) => this.map[i + point.dx][j + point.dy]).reduce((prev, curr, i, array) => curr + prev, 0) + 1;
        if (probableShip > 4) return false;
        if (this.ships[probableShip] + 1 + probableShip > 5) return false;
        return true;
    } 

    private recalcShips() {
        this.ships = { '1': 0, '2': 0, '3': 0, '4': 0 };
        for (let i = 0; i < this.map.length; i++) {
            for (let j = 0; j < this.map.length; j++) {
                if (this.map[i][j]) {
                    this.ships[this.map[i][j]]++;
                }
            }
        }
    }
}

File: LiBattleship.Shared/Models/Field.cs
using LiBattleship.Shared.Enums;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace LiBattleship.Shared.Models
{
    public class Field
    {
        enum ShipState
        {
            Unknown = 0,
            Hitted = 16,
            Killed = 32
        }

        /* *
         * CELL BIT STRUCTURE:
         * 
         * 0
         * XXXYZZ
         * 
         * X - SHIP SIZE FIELD (000 if no ship here);
         * Y - HIT STATE;
         * Z - SHIP STATE (SETTED ONLY IF Y = 1);
         *      00 - UNKNOWN OR NO SHIP HERE
         *      01 - HITTED
         *      10 - KILLED
         * */
        
        private readonly int[][] _inner;

        public Field(int[][] field)
        {
            _inner = field;
        }

        private bool GetHitState(int cell)
        {
            return (cell & 8) != 0;
        }

        private ShipState GetShipState(int cell)
        {
            return (ShipState)(cell & 48);
        }

        private int GetShipSize(int cell)
        {
            return cell & 7;
        }

        private void SetHitState(int i, int j, bool state)
        {
            _inner[i][j] = (_inner[i][j] & ~8) | (Convert.ToByte(state) << 3);
        }

        private void SetShipState(int i, int j, ShipState state)
        {
            _inner[i][j] = (_inner[i][j] & ~48) | Convert.ToByte(state);
        }

        private void SetShipSize(int[][] array, int i, int j, int size)
        {
            array[i][j] = (array[i][j] & ~7) | (size & 7);
        }

        public MoveResult MakeMove(int x, int y)
        {
            if (!IsValidMove(x, y)) return MoveResult.IncorrectMove;
            SetHitState(x, y, true);

            if (GetShipSize(_inner[x][y]) != 0)
            {
                SetShipState(x, y, ShipState.Hitted);
                CheckKilledAndSet(x, y);
                if (IsShipCountValid(GetRawData(true))) {
                    return MoveResult.GameFinished;
                }
                return MoveResult.Hit;
            }
            return MoveResult.NoHit;
        }

        private void CheckKilledAndSet(int x, int y)
        {
            var shipCells = new List<(int x, int y)>() { (x, y) };
            var nearestPoints = this.GetNearestPoints(x, y);

            foreach (var (dx, dy) in nearestPoints)
            {
                var i = x;
                var j = y;
                while (true)
                {
                    if (i + dx < _inner.Length && i + dx >= 0
                        && j + dy < _inner.Length && j + dy >= 0 && GetShipSize(_inner[i + dx][j + dy]) > 0)
                    {
                        i = i + dx;
                        j = j + dy;
                    }
                    else break;
                    if (dx == 0 && dy == 0) break;
                    else shipCells.Add((i, j));
                }
            };

            if (shipCells.All(cell => GetShipState(_inner[cell.x][cell.y]) == ShipState.Hitted)) {
                shipCells.ForEach(cell => SetShipState(cell.x, cell.y, ShipState.Killed));
            }
        }

        public int[][] GetRawData(bool masked)
        {
            var result = new int[10][];
            for (int i = 0; i < 10; i++)
            {
                result[i] = new int[10];
                for (int j = 0; j < 10; j++)
                {
                    result[i][j] = _inner[i][j];
                    if (masked && GetShipState(_inner[i][j]) != ShipState.Killed)
                    {
                        SetShipSize(result, i, j, 0);
                    }
                }
            }
            return result;
        }

        private IEnumerable<(int dx, int dy)> GetNearestPoints(int x, int y)
        {
            var result = new List<(int dx, int dy)>();
            if (GetShipSize(_inner[x][y]) > 0) { result.Add((0, 0)); }
            if (x > 0 && GetShipSize(_inner[x - 1][y]) > 0) { result.Add((-1, 0)); }
            if (y < _inner.Length - 1 && GetShipSize(_inner[x][y + 1]) > 0) { result.Add((0, 1)); }
            if (y > 0 && GetShipSize(_inner[x][y - 1]) > 0) { result.Add((0, -1)); }
            if (x < _inner.Length - 1 && GetShipSize(_inner[x + 1][y]) > 0) { result.Add((1, 0)); }
            return result;
        }

        private bool IsValidMove(int i, int j)
        {
            if (GetHitState(_inner[i][j])) return false;
            if (i > 0 && j > 0 && GetShipState(_inner[i - 1][j - 1]) != ShipState.Unknown) return false;
            if (i > 0 && j < _inner.Length - 1 && GetShipState(_inner[i - 1][j + 1]) != ShipState.Unknown) return false;
            if (j > 0 && i < _inner.Length - 1 && GetShipState(_inner[i + 1][j - 1]) != ShipState.Unknown) return false;
            if (j < _inner.Length - 1 && i < _inner.Length - 1 && GetShipState(_inner[i + 1][j + 1]) != ShipState.Unknown) return false;
            //var probableShip = this.getNearestPoints(i, j).Select((point) => _inner[i + point.dx][j + point.dy]).Sum() + 1;
            //if (probableShip > 4) return false;
            return true;
        }


        private bool IsShipCountValid(int[][] field)
        {
            var ships = new Dictionary<int, int>() {
                { 1, 0 },
                { 2, 0 },
                { 3, 0 },
                { 4, 0 }
            };

            for (int i = 0; i < field.Length; i++) {
                for (int j = 0; j < field[i].Length; j++) {
                    var shipSize = GetShipSize(field[i][j]);
                    if (shipSize > 0) ships[shipSize]++;
                }
            }
            foreach (var value in ships) {
                if (value.Value / value.Key != 5 - value.Key) return false;
            }

            return true;
        }
    }
}
